resources/lib/youtube_plugin/kodion/impl/abstract_settings.py

```diff
@@ -193,62 +209,60 @@ def get_location(self):
+        self.set_string(SETTINGS.LOCATION, value)
 
     def get_location_radius(self):
-        return ''.join([str(self.get_int(constants.setting.LOCATION_RADIUS, 500)), 'km'])
+        return ''.join([str(self.get_int(SETTINGS.LOCATION_RADIUS, 500)), 'km'])
 
     def get_play_count_min_percent(self):
-        return self.get_int(constants.setting.PLAY_COUNT_MIN_PERCENT, 0)
-
-    def use_playback_history(self):
-        return self.get_bool(constants.setting.USE_PLAYBACK_HISTORY, False)
-
-    @staticmethod
-    def __get_mpd_quality_map():
-        return {
-            0: 240,
-            1: 360,
-            2: 480,
-            3: 720,
-            4: 1080,
-            5: 1440,
-            6: 2160,
-            7: 4320,
-            8: 'mp4',
-            9: 'webm'
-        }
-
-    def get_mpd_quality(self):
-        quality_map = self.__get_mpd_quality_map()
-        quality_enum = self.get_int(constants.setting.MPD_QUALITY_SELECTION, 8)
-        return quality_map.get(quality_enum, 'mp4')
-
-    def mpd_video_qualities(self):
-        if not self.use_dash_videos():
+        return self.get_int(SETTINGS.PLAY_COUNT_MIN_PERCENT, 0)
+
+    def use_local_history(self):
+        return self.get_bool(SETTINGS.USE_LOCAL_HISTORY, False)
+
+    def use_remote_history(self):
+        return self.get_bool(SETTINGS.USE_REMOTE_HISTORY, False)
+
+    # Selections based on max width and min height at common (utra-)wide aspect ratios
+    _QUALITY_SELECTIONS = {                                                 # Setting | Resolution
+        7:   {'width': 7680, 'height': 3148, 'label': '4320p{0} (8K){1}'},  #   7     |   4320p 8K
+        6:   {'width': 3840, 'height': 1080, 'label': '2160p{0} (4K){1}'},  #   6     |   2160p 4K
+        5:   {'width': 2560, 'height': 984, 'label': '1440p{0} (QHD){1}'},  #   5     |   1440p 2.5K / QHD
+        4.1: {'width': 2048, 'height': 858, 'label': '1152p{0} (2K){1}'},   #   N/A   |   1152p 2K / QWXGA
+        4:   {'width': 1920, 'height': 787, 'label': '1080p{0} (FHD){1}'},  #   4     |   1080p FHD
+        3:   {'width': 1280, 'height': 525, 'label': '720p{0} (HD){1}'},    #   3     |   720p  HD
+        2:   {'width': 854, 'height': 350, 'label': '480p{0}{1}'},          #   2     |   480p
+        1:   {'width': 640, 'height': 263, 'label': '360p{0}{1}'},          #   1     |   360p
+        0:   {'width': 426, 'height': 175, 'label': '240p{0}{1}'},          #   0     |   240p
+        -1:  {'width': 256, 'height': 105, 'label': '144p{0}{1}'},          #   N/A   |   144p
+        -2:  {'width': 0, 'height': 0, 'label': '{2}p{0}{1}'},              #   N/A   |   Custom
+    }
+
+    def get_mpd_video_qualities(self):
+        if not self.use_mpd_videos():
             return []
+        selected = self.get_int(SETTINGS.MPD_QUALITY_SELECTION, 4)
+        return [quality for key, quality in self._QUALITY_SELECTIONS.items()
+                if selected >= key]
```

Comment:
dict iteration order matching insertion order is a Python 3.7+ feature. The order of the quality selections is important when sorting and selecting the available adaptive streams - if it is not in order then the quality comparisons may fail and the streams will be skipped.

This may fail in Kodi 19+ as well depending on the version of Python used, so I will update this in master to be:
```
        return [quality
                for key, quality in sorted(
                    self._QUALITY_SELECTIONS.items(),
                    reverse=True
                )
                if selected >= key]
```

For Kodi 18 you can additionally change `self._QUALITY_SELECTIONS.items()` to be `self._QUALITY_SELECTIONS.viewitems()`